doc/source/devel/biaps/biap_0009.rst

```diff
@@ -0,0 +1,309 @@
+
+        def get_triangles(self, name=None):
+            """ Mx3 array of indices into coordinate table """
+
+        def get_mesh(self, name=None):
+            return self.get_coords(name=name), self.get_triangles(name=name)
+
+        def get_names(self):
+            """ List of surface names that can be passed to
+            ``get_{coords,triangles,mesh}``
+            """
+
+        def decimate(self, *, n_coords=None, ratio=None):
+            """ Return a SurfaceHeader with a smaller number of vertices that
+            preserves the geometry of the original """
+            # To be overridden when a format provides optimization opportunities
+
+        def load_vertex_data(self, filename):
+            """ Return a SurfaceImage with data corresponding to each vertex """
+
+        def load_face_data(self, filename):
+            """ Return a SurfaceImage with data corresponding to each face """
+
+
+    class SurfaceHeader(FileBasedHeader):
+        ...
+
+
+    class SurfaceImage(FileBasedImage):
+        @property
+        def header(self):
+            """ A SurfaceHeader """
+
+        @property
+        def geometry(self):
+            """ A SurfaceGeometry or None """
+
+        @property
+        def dataobj(self):
+            """ An ndarray or ArrayProxy with one of the following properties:
+            
+            1) self.dataobj.shape[0] == self.header.ncoords
+            2) self.dataobj.shape[0] == self.header.nfaces
+            """
+
+        def load_geometry(self, pathlike):
+            """ Specify a geometry for a data-only image """
+
+
+To enable a similar interface to raveled voxel data:
+
+.. code-block:: python
+
+    class VolumeGeometry(Geometry):
+        _affine  # Or _affines, if we want multiple, e.g. qform, sform
+        _shape
+        _ijk_coords
+
+        def n_coords(self):
+            """ Number of coordinates """
+
+        def get_coords(self):
+            """ Nx3 array of coordinates in RAS+ space """
+
+
+Here we present common use cases:
+
+
+Modeling
+========
+
+.. code-block:: python
+
+    from nilearn.glm.first_level import make_first_level_design_matrix, run_glm
+
+    bold = SurfaceImage.from_filename("/data/func/hemi-L_bold.func.gii")
+    dm = make_first_level_design_matrix(...)
+    labels, results = run_glm(bold.get_data(), dm)
+    betas = bold.__class__(results["betas"], bold.header)
+    betas.to_filename("/data/stats/hemi-L_betas.mgz")
+
+Data images have their own metadata apart from geometry that needs handling in a header:
+
+* Axis labels (time series, labels, tensors)
+* Data dtype
+
+
+Smoothing
+=========
+
+.. code-block:: python
+
+    bold = SurfaceImage.from_filename("/data/func/hemi-L_bold.func.gii")
+    bold.load_geometry("/data/anat/hemi-L_midthickness.surf.gii")
+    # Not implementing networkx weighted graph here, so assume we have a method
+    graph = bold.geometry.get_graph()
+    distances = distance_matrix(graph)  # n_coords x n_coords matrix
```

Comment:
This is maybe going to be harder than we think:
```
>>> rr, tris = nib.freesurfer.read_geometry('/mnt/bakraid/data/structurals/fsaverage/surf/lh.white')
>>> from mne.surface import mesh_dist
>>> adjacency = mesh_dist(tris, rr)
>>> t0 = time.time(); dist_matrix = dijkstra(adjacency, directed=False, limit=10.); print(time.time() - t0)
...
numpy.core._exceptions.MemoryError: Unable to allocate 200. GiB for an array with shape (163842, 163842) and data type float64
```
So we'll need to improve SciPy's dijkstra to support `sparse` outputs start...

If we don't use the distance-based method and instead use the "smudging" operation by pushing data to neighboring vertices using a number of steps, this process is avoided. I guess this is a bit of an actual implementation detail, but we should think about these two methods as options with different tradeoffs.

Reply:
Yes, full pairwise distances is painfully naive; not sure if iterated smudging is equivalent to distance-based (up to accumulated floating point error) but seems reasonable.

Mostly as a bookmark to myself, it might be worth looking into https://github.com/gilShamai/Fast-Pairwise-Geodesic-Distance-Computation to see if they have any good ideas.